sql/core/src/main/scala/org/apache/spark/sql/execution/GenerateExec.scala

```diff
@@ -226,14 +226,17 @@ case class GenerateExec(
       "0"
     }
     val numOutput = metricTerm(ctx, "numOutputRows")
+    val requiredInput = input.zip(child.output).filter {
+      case(_, output) => parent.inputSet.contains(output)
+    }.map(_._1) ++ position ++ values
```

Comment:
Shall we do this in `doConsume`, and pass `requiredInput` to both `codeGenCollection` and `codeGenTraversableOnce`?